external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaWriteTask.scala

```diff
@@ -0,0 +1,119 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.kafka010
+
+import java.{util => ju}
+
+import org.apache.kafka.clients.producer._
+import org.apache.kafka.common.serialization.ByteArraySerializer
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Attribute, Cast, Literal, UnsafeProjection}
+import org.apache.spark.sql.types.{BinaryType, StringType}
+import org.apache.spark.unsafe.types.UTF8String
+
+/**
+ * A simple trait for writing out data in a single Spark task, without any concerns about how
+ * to commit or abort tasks. Exceptions thrown by the implementation of this class will
+ * automatically trigger task aborts.
+ */
+private[kafka010] class KafkaWriteTask(
+    producerConfiguration: ju.Map[String, Object],
+    inputSchema: Seq[Attribute],
+    topic: Option[String]) {
+  // used to synchronize with Kafka callbacks
+  @volatile var failedWrite: Exception = null
+  val topicExpression =
+    if (topic.isDefined) {
+      // topic option overrides topic field
+      Literal(UTF8String.fromString(topic.get), StringType)
+    } else {
+    inputSchema.find(p =>
+      p.name == KafkaWriter.TOPIC_ATTRIBUTE_NAME).getOrElse(
+        throw new IllegalStateException(s"topic option required when no " +
+          s"'${KafkaWriter.TOPIC_ATTRIBUTE_NAME}' attribute is present"))
+    }
+  val keyExpression = inputSchema.find(p =>
+    p.name == KafkaWriter.KEY_ATTRIBUTE_NAME).getOrElse(
+      Literal(null, BinaryType)
+    )
+  keyExpression.dataType match {
+    case StringType | BinaryType => // good
+    case t =>
+      throw new IllegalStateException(s"${KafkaWriter.KEY_ATTRIBUTE_NAME} " +
+        s"attribute unsupported type $t")
+  }
+  val valueExpression = inputSchema.find(p =>
+    p.name == KafkaWriter.VALUE_ATTRIBUTE_NAME).getOrElse(
+      throw new IllegalStateException(s"Required attribute " +
+        s"'${KafkaWriter.VALUE_ATTRIBUTE_NAME}' not found")
+    )
+  valueExpression.dataType match {
+    case StringType | BinaryType => // good
+    case t =>
+      throw new IllegalStateException(s"${KafkaWriter.VALUE_ATTRIBUTE_NAME} " +
+        s"attribute unsupported type $t")
+  }
+  val projection = UnsafeProjection.create(
+    Seq(topicExpression, Cast(keyExpression, BinaryType),
+      Cast(valueExpression, BinaryType)), inputSchema)
+
+  // Create a Kafka Producer
+  producerConfiguration.put("key.serializer", classOf[ByteArraySerializer].getName)
+  producerConfiguration.put("value.serializer", classOf[ByteArraySerializer].getName)
+  val producer = new KafkaProducer[Array[Byte], Array[Byte]](producerConfiguration)
```

Comment:
In addition, i think we should not create producer until its used, since the creation of KafkaWriterTask is outside the try. If code is moved around, it may so happen that an error in the KafkaWriterTask constructor *after the producer* is created will end up escaping the try-catch and the producer would never close.